### PR TITLE
Add tests for bool vector and fix implementation

### PR DESCRIPTION
**Pull request by shariarriday**
I have added a serialization/deserialization test for binary serializer. 

But, the previous implementation had a bug which caused the specialized implementation to not be executed. Due to the nature of the bug, I could not find any solution circumventing the ABI breaking changes. 
I think that we can plan to merge it in for 2.0. I am creating a draft pull request for this reason.

### CODE REVIEW

File: libcaf_core/caf/serialization.test.cpp [hunk 1]
```diff
@@ -729,11 +746,13 @@ OUTLINE("serializing and then deserializing primitive values") {
     | binary_serializer   | u32    | 123456        |
     | binary_serializer   | u64    | 123456789     |
     | binary_serializer   | ld     | 123.5         |
+    | binary_serializer   | f      | 3.14          |
     | binary_serializer   | real   | 12.5          |
     | binary_serializer   | string | Hello, world! |
     | binary_serializer   | u16str | Hello, world! |
     | binary_serializer   | u32str | Hello, world! |
     | binary_serializer   | vector | 1, 42, -31    |
+    | binary_serializer   | vector_bool | true,true,false,false,true,true,true,false |
```
Comment (Neverlord) [inline libcaf_core/caf/serialization.test.cpp:755]: That's a bit unwindy, abbreviating it indents nicer and is less verbose:

```suggestion
    | binary_serializer   | vbool | true,true,false,false,true,true,true,false |
```

Is there a particular reason why this has 8 values? Could just have two to avoid this overly long line.

**Review by Neverlord — CHANGES_REQUESTED** [review]
It's only an ABI break because it removes the `value(std::vector<bool>& x)` overload. Bring that back and simply call that from the new `list(value(std::vector<bool>& x)` overload. Then we can merge it right away.

File: libcaf_core/caf/serialization.test.cpp (hunk 1 above)
Comment (shariarriday) [reply libcaf_core/caf/serialization.test.cpp:755]: The serializer and deserializer use a switch case to serialize/deserialize an array into a single byte. I have added 8 entries to ensure that all conditions are being handled properly. Please let me know if it will be okay to skip it.

**Review by Neverlord — CHANGES_REQUESTED** [review]
Sorry, I've realized that we cannot pull this in now. It's not an ABI break, but it changes the wire format. We don't want to change our data representation on the network in a minor release.

Please split this up into one PR for the `serialization.test.cpp` additions and one PR for the fix (that will go into a v2 branch).

**Comment by shariarriday** [comment]
> Sorry, I've realized that we cannot pull this in now. It's not an ABI break, but it changes the wire format. We don't want to change our data representation on the network in a minor release.
> 
> Please split this up into one PR for the `serialization.test.cpp` additions and one PR for the fix (that will go into a v2 branch).

I am not sure splitting up the tests would be beneficial. The added test will not hit the specialization for serializing boolean vectors and it will just execute the code of serializing any vectors. Maybe it will be best to include the test and code changes in a single PR for 2.x . Please let me know if you think otherwise.

**Comment by Neverlord** [comment]
I think having the tests separately pulled in now is still worth it. The non-binary inspectors do a have a separate code path for `vector<bool>` that currently isn't tested. Also:

> then add a separate test for the binary serialization that consumes bool vectors of various sizes (at the very least the edge cases, i.e., 0, 1, 8, 9).

This test should still go in. For now, this test can call the `value(...)` overload for `vector<bool>` to get the coverage on that code path, ideally with a TODO reminder to use `apply` instead once we pull in the other change.

File: libcaf_core/caf/serialization.test.cpp [hunk 2]
```diff
@@ -861,6 +884,55 @@ OUTLINE("serializing a type that initializes members to a non-empty state") {
...
+SCENARIO("binary serializer and deserializer handle vectors of booleans") {
+  GIVEN("a binary serializer") {
+    auto sink = binary_serializer_wrapper{sys};
+    WHEN("serializing a vector of booleans with 8 values") {
+      auto val = std::vector<bool>{true,  false, true, true,
+                                   false, false, true, false};
+      check(sink.sink.value(val));
+      THEN("deserializing the result produces the value again") {
+        auto source = binary_deserializer{sys, sink.buffer};
+        std::vector<bool> copy = {};
```
Comment (Neverlord) [inline libcaf_core/caf/serialization.test.cpp:897]: For keeping a consistent style:

```suggestion
        auto copy = std::vector<bool>{};
```

Same in the occurrences below.

**Review by Neverlord — CHANGES_REQUESTED** [review]
Just one style nitpick. 👍
